packages/core/src/controllers/ChainController.ts

```diff
@@ -117,41 +113,49 @@ export const ChainController = {
     })
   },
 
-  setMultiChainEnabled(multiChain: boolean) {
-    state.multiChainEnabled = multiChain
-  },
-
-  setChainNetworkData(chain: Chain | undefined, props: Partial<NetworkControllerState>) {
+  setChainNetworkData(
+    chain: Chain | undefined,
+    props: Partial<NetworkControllerState>,
+    replaceState = false
+  ) {
     if (!chain) {
       throw new Error('Chain is required to update chain network data')
     }
 
     const chainAdapter = state.chains.get(chain)
 
     if (chainAdapter) {
-      chainAdapter.networkState = {
+      chainAdapter.networkState = ref({
         ...chainAdapter.networkState,
         ...props
-      } as NetworkControllerState
-      state.chains.set(chain, chainAdapter)
-      NetworkController.replaceState(chainAdapter.networkState)
+      } as NetworkControllerState)
+      state.chains.set(chain, ref(chainAdapter))
+      if (replaceState || state.chains.size === 1 || state.activeChain === chain) {
+        NetworkController.replaceState(chainAdapter.networkState)
+      }
     }
   },
 
-  setChainAccountData(chain: Chain | undefined, accountProps: Partial<AccountControllerState>) {
+  setChainAccountData(
+    chain: Chain | undefined,
+    accountProps: Partial<AccountControllerState>,
+    replaceState = true
+  ) {
     if (!chain) {
       throw new Error('Chain is required to update chain account data')
     }
 
     const chainAdapter = state.chains.get(chain)
 
     if (chainAdapter) {
-      chainAdapter.accountState = {
+      chainAdapter.accountState = ref({
```

Comment:
Why ref

Reply:
When setting not primitive values (objects, arrays), valtio suggests using `ref`